### DISCUSSION TITLE
Thorlabs XYZ piezo stage

### DISCUSSION THREAD

**Discussion by RuairidhCunningham** [Ulm-IQO/qudi-core#88] [Q&A]
Hi there folks,

I am a PhD student currently working on an experimental set-up that uses Qudi to control and integrate all the hardware such as an Ni series DAQ card, we also have a piezo stage from Thorlabs, specifically this model here:
https://www.thorlabs.com/NewGroupPage9.cfm?ObjectGroup_ID=2386

Controlled through these controller modules:
https://www.thorlabs.com/newgrouppage9.cfm?objectgroup_id=2420

That are then connected to the PC.  I have found the aptmotor.py hardware file in the qudi folder and was wondering if 1)it would work with these stepper motor controllers and 2) what the config file would look like for adding the aptmotor.py hardware? 

This is to be used to create a confocal scan across the XY of a sample.  Also is the XYZ controller's ok to be connected to the PC via USB or does it require to be connected to our BNC box and DAQ card as well?

Thanks for your time and all the best,
Ruairidh

**Comment by timoML** [comment]
Hi,

there is a very similar question, already [answered](https://github.com/Ulm-IQO/qudi-iqo-modules/discussions/104) in the iqo-modules subrepo. In short:
- The aptmtotor.py file is not fulfilling the interface required for the (confocal) scanning probe toolchain. 
- You could try implementing a new hardware file for your Thorlabs controllers that implements the scanning_probe_interface.
- Easier (if supported by your hardware) could be to use the "IQO way" of generating analogue outputs from an nicard and input this to your Piezo controllers.

We can try helping you with the first steps, if you decdied to develop something new.

**Reply by RuairidhCunningham** [reply]
Hi timoML,

I'm going for the "IQO way" which might offer a solution since writing a hardware file might be a little outside of my expertise!

We have our DAQ card (PCIe 6323) connected to our BNC-2110 connected to our KST201 Thorlabs piezo controllers.  We're able to make the Controller's move the stage (hooray!) using a Voltage Output from the DAQ card, however when trying to perform the confocal scan it doesn't seem to work.  The controller is triggered to move by the rising edge of a voltage above 2.4V.

From what I can piece together from the Hardware file for the national_instruments_x_series.py it almost looks like a Voltage is assigned to each pixel co-ordinate and then this voltage is supplied to command the stage to move to the corresponding position.  Obviously, this wouldn't work with our set-up since our controller's are only triggered by a single specific voltage.

I suppose my question boils down to, is it possible to perform this confocal scan using a single pulsed, fixed Voltage from the DAQ card in order to give the move command to the controller?

Apologies if any of this doesn't make sense or is unclear.

Thanks for your quick response!

**Reply by timoML** [reply]
I'm not familiar with your hardware. But assuming that you currently move by digitally prodiving a position and a subsequent analogue trigger, I don't think it's worth adapting our [NiScanningProbeInterfuse](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/main/src/qudi/hardware/interfuse/ni_scanning_probe_interfuse.py#L40)

You're right, we're providing positions by encoding them on the analogue outputs of the nicard. So every function in NiScanningProbeInterfuse that implements the [ScanninProbeInterface](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/212103937862ffe0744661045004858d728ad173/src/qudi/interface/scanning_probe_interface.py#L29) is assuming this way of output coding. If this is different for you, it seems more straightforward to me to write a new file that implements the interface with functions that issue (digital) commands to your controller.

**Comment by Neverhorst** [comment]
Duplicate of discussion https://github.com/Ulm-IQO/qudi-iqo-modules/discussions/104
